src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -0,0 +1,419 @@
+                        break;
+
+                    case QUIC_CONNECTION_EVENT.STREAMS_AVAILABLE:
+                        {
+                            status = HandleEventStreamsAvailable(
+                                connectionEvent);
+                        }
+                        break;
+
+                    default:
+                        break;
+                }
+            }
+            catch (Exception)
+            {
+                // TODO we may want to either add a debug assert here or return specific error codes
+                // based on the exception caught.
+                return MsQuicConstants.InternalError;
+            }
+
+            return status;
+        }
+
+        private uint HandleEventConnected(ConnectionEvent connectionEvent)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            SOCKADDR_INET inetAddress = MsQuicParameterHelpers.GetINetParam(_api, _ptr, (uint)QUIC_PARAM_LEVEL.CONNECTION, (uint)QUIC_PARAM_CONN.LOCAL_ADDRESS);
+            _localEndPoint = MsQuicAddressHelpers.INetToIPEndPoint(inetAddress);
+
+            _connected = true;
+            // I don't believe we need to lock here because
+            // handle event connected will not be called at the same time as
+            // handle event shutdown initiated by transport
+            _connectTcs.Complete(MsQuicConstants.Success);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventShutdownInitiatedByTransport(ConnectionEvent connectionEvent)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            if (!_connected)
+            {
+                _connectTcs.CompleteException(new IOException("Connection has been shutdown."));
+            }
+
+            _acceptQueue.Writer.Complete();
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventShutdownInitiatedByPeer(ConnectionEvent connectionEvent)
+        {
+            _acceptQueue.Writer.Complete();
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventShutdownComplete(ConnectionEvent connectionEvent)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            _shutdownTcs.Complete(MsQuicConstants.Success);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventNewStream(ConnectionEvent connectionEvent)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            MsQuicStream msQuicStream = new MsQuicStream(_api, this, connectionEvent.StreamFlags, connectionEvent.Data.NewStream.Stream, inbound: true);
+
+            _acceptQueue.Writer.TryWrite(msQuicStream);
```

Comment:
Backlog handling here? How does this work?

Reply:
There is no backlog for accepting streams from connections; it's handled by MsQuic. When we create the MsQuicSession today, we set the max bidirectional and unidirectional streams on a connection. MsQuic is guaranteed to never exceed these limits.

Reply:
The only thing that I find a bit confusing is how these limits are negotiated between peers. @nibanks if I set the bidirectional and unidirectional stream limit on the session, when a connection is accepted, those limits get applied to the peer, correct?

Reply:
The max streams setting affects all open streams, so it's not quite the same as backlog.

I'm wondering if there's some scenario where you might want to have a large (or infinite) value for max streams, but a smaller value for backlog. Like say you have lots of long-lived streams, but you still want to prevent against a flood of streams being created all at once. I dunno, but it seems worth considering.

But we don't need to solve this for this PR.

Reply:
Yeah, I'm not convinced we need it for streams. For connections it makes sense though.